File: mpi/code/c/solution/helloworld_mpi3b.c
/*
 * first example with a deadlock
 *
 * take care that the deadlock 'disappear' for small message length
 *
 * Try to use N=1000 (no deadlock with OpenMPI) and N=10000 (deadlock)
 *
 * Try to use mpirun option '--mca btl_vader_eager_limit 8192' to increase
 * the eager limit
 */

#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

static const int N=10000;

int main(int argc, char* argv[])
{

  int nbTask;
  int myRank;

  /* send / recv data array */
  int dataS[N], dataR[N];
  int i;

  MPI_Status status;

  int tag1 = 1;
  int tag2 = 2;

  MPI_Init(&argc, &argv);

  MPI_Comm_size(MPI_COMM_WORLD, &nbTask);
  MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

  /* data to send initialization */
  for (i=0; i<N; i++) {
    dataS[i] = (myRank+1)*i;
    dataR[i] = 0;
  }

  printf("I am task %d out of %d\n",myRank,nbTask);

  if (myRank == 0) {

    printf("[Task %d]: I'm sending data to rank 1\n",myRank);
    MPI_Send(&(dataS[0]), N, MPI_INT, 1, tag1, MPI_COMM_WORLD);

    MPI_Recv(&(dataR[0]), N, MPI_INT, 1, tag2, MPI_COMM_WORLD, &status);
    printf("[Task %d] I received data from task 1\n",myRank);

  } else if (myRank == 1) {

    printf("[Task %d]: I'm sending data to rank 0\n",myRank);
    MPI_Send(&(dataS[0]), N, MPI_INT, 0, tag2, MPI_COMM_WORLD);

    MPI_Recv(&(dataR[0]), N, MPI_INT, 0, tag1, MPI_COMM_WORLD, &status);
    printf("[Task %d] I received data from task 0\n",myRank);
  }

  MPI_Finalize();

  return 0;

}

File: mpi/code/fortran/solution/helloworld_mpi3b.f90
!
! first example with a deadlock 
!
! take care that the deadlock 'disappear' for small message length
!
! Try to use N=1000 (no deadlock with OpenMPI) and N=10000 (deadlock)
!


program helloworld_mpi3

  use mpi

  implicit none

  integer   :: nbTask, myRank, ierr
  integer, dimension(MPI_STATUS_SIZE)         :: status

  integer, parameter :: N=10000
  integer :: dataToSend(N), dataRecv(N)

  call MPI_Init(ierr)

  call MPI_COMM_SIZE(MPI_COMM_WORLD, nbTask, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)

  write (*,*) 'I am task', myRank, 'out of',nbTask

  if (myRank == 0) then

     dataToSend = 33

     write (*,*) "[Task ",myRank,"] I'm sending data to rank 1", dataToSend
     call MPI_SEND(dataToSend, N, MPI_INTEGER, 1, 100, MPI_COMM_WORLD, ierr)

     call MPI_RECV(dataRecv,   N, MPI_INTEGER, 1, 100, MPI_COMM_WORLD, status, ierr)
     write (*,*) "[Task ",myRank,"] I received data  from task 0",dataRecv

  else if (myRank == 1) then

     dataToSend = 34

     write (*,*) "[Task ",myRank,"] I'm sending data to rank 0"
     call MPI_SEND(dataToSend, N, MPI_INTEGER, 0, 100, MPI_COMM_WORLD, ierr)

     call MPI_RECV(dataRecv,   N, MPI_INTEGER, 0, 100, MPI_COMM_WORLD, status, ierr)
     write (*,*) "[Task ",myRank,"] I received data  from task 0"


  end if

  call MPI_Finalize(ierr)

end program helloworld_mpi3
